test/test.downscopedclient.ts

```diff
@@ -0,0 +1,406 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as assert from 'assert';
+import {describe, it, before, after, beforeEach, afterEach} from 'mocha';
+import * as nock from 'nock';
+import * as sinon from 'sinon';
+
+import {StsSuccessfulResponse} from '../src/auth/stscredentials';
+import {DownscopedClient} from '../src/auth/downscopedclient';
+import {AuthClient} from '../src/auth/authclient';
+import {mockStsBetaTokenExchange} from './externalclienthelper';
+import {GoogleAuth} from '../src';
+
+nock.disableNetConnect();
+
+interface SampleResponse {
+  foo: string;
+  bar: number;
+}
+
+describe('DownscopedClient', () => {
+  let clock: sinon.SinonFakeTimers;
+
+  const auth = new GoogleAuth({
+    keyFilename: './test/fixtures/private.json',
+    scopes: 'https://www.googleapis.com/auth/cloud-platform',
+  });
+  let client: AuthClient;
+
+  const ONE_HOUR_IN_SECS = 3600;
+  const testAvailableResource =
+    '//storage.googleapis.com/projects/_/buckets/bucket-123';
+  const testAvailablePermission1 = 'inRole:roles/storage.objectViewer';
+  const testAvailablePermission2 = 'inRole:roles/storage.objectAdmin';
+  const testAvailabilityConditionExpression =
+    "resource.name.startsWith('projects/_/buckets/bucket-123/objects/prefix')";
+  const testAvailabilityConditionTitle = 'Test availability condition title.';
+  const testAvailabilityConditionDescription =
+    'Test availability condition description.';
+  const testClientAccessBoundary = {
+    accessBoundary: {
+      accessBoundaryRules: [
+        {
+          availableResource: testAvailableResource,
+          availablePermissions: [testAvailablePermission1],
+          availabilityCondition: {
+            expression: testAvailabilityConditionExpression,
+          },
+        },
+      ],
+    },
+  };
+  const stsSuccessfulResponse: StsSuccessfulResponse = {
+    access_token: 'DOWNSCOPED_CLIENT_ACCESS_TOKEN_0',
+    issued_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+    token_type: 'Bearer',
+    expires_in: ONE_HOUR_IN_SECS,
+    scope: 'scope1 scope2',
```

Comment:
Let's remove the scope field. This is not [documented](https://cloud.google.com/iam/docs/downscoping-short-lived-credentials#create-credential) in the response.